clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java

```diff
@@ -523,52 +531,66 @@ public void fireOnAssignment(Set<TopicPartition> assignment) {
             this.nextAllowedRetryTimeMs = null;
         }
 
-        private void reset(OffsetResetStrategy strategy) {
+        private synchronized void reset(OffsetResetStrategy strategy) {
```

Comment:
We still have a few places where we are accessing the internal fields directly without the lock (e.g. `updateHighWatermark`). Do you think it's worthwhile protecting all accesses?

As a matter of fact, I had resolved another synchronization problem in https://issues.apache.org/jira/browse/KAFKA-7831. However, I had resisted making `SubscriptionState` itself synchronized because of the presence of `SubscriptionState.Listener`. But now I'm wondering if that would be the simplest/safest option. In fact, the only use of the listener is this call below inside `Fetcher`. 

```java
    @Override
    public void onAssignment(Set<TopicPartition> assignment) {
        sensors.updatePartitionLagAndLeadSensors(assignment);
    }
```

A simple alternative would be to update the sensors whenever `sendFetches` is called. For example, we could let SubscriptionState track a simple counter which is incremented after every new assignment. Then it would be easy to check for any changes. Thoughts?

Reply:
As I see `updateHighWatermark` isn't used across threads in the client itself so I'm not sure. It's definitely low effort to update but at this point I'm not sure about the gains. For future proofing and because it may look nicer I think it can be updated. (And with this the other fields, like `logStartOffset`, `lastStableOffset` and `paused`.)

Regarding `SubscriptionState`: generally I think it is a good idea as it stores and manages state and accessed from multiple threads., although I have limited knowledge about its usage. I really like how `ProducerMetadata` and `Metadata` are written for instance as they are reasonably sized, encapsulate the necessary data and checks on them. You instantly know that what data needs to be handled together. 
I only have one concern with `SubscriptionState` which basically comes from my blind spots is that we might synchronize some methods which needn't to be synchronized or we introduce some starving/deadlocking situation. 
On the other hand looking at some methods (`groupSubscribe`, `changeSubscription`) lets me think that we want to handle `subscription` and `groupSubscription` together so both are up to date, however only groupSubscription is concurrent, so probably it is used from multiple threads. I got a bit lost in the method calls while figuring out how it is used exactly but it seems like a code smell at least.
Also we can say that we don't provide synchronization for `listeners` (like `ConsumerNetworkClient`), it seems pretty independent of the other data in there anyway. Your alternative is also good, honestly I'd be good with either :).

Reply:
I think we should try to optimize for the simplest locking that is possible even if it is unnecessarily coarse. `Metadata` is a good example where it might be possible to do finer locking, but it's not really worth the effort. Otherwise we need to put a lot of effort into understanding current usage and making sure we don't break locking expectations in the future. There is really only the background heartbeat thread which is contending for these locks, so I'm not too worried about lock contention in general.

> Also we can say that we don't provide synchronization for listeners (like ConsumerNetworkClient), it seems pretty independent of the other data in there anyway. Your alternative is also good, honestly I'd be good with either :).

That's fair. As long as we can invoke listeners outside of the lock, it's probably ok. I think I'd still like to get rid of the listener because it makes the code harder to understand, but perhaps we can treat this as a separate issue.

Reply:
FYI: #6559